### DISCUSSION TITLE
Jumper R900 support

### DISCUSSION THREAD

**Discussion by FelixSchreiber** [ExpressLRS/ExpressLRS#1114] [Ideas]
Now as there is #998 PWM Servo Output Support available - what about adding support for Jumper R900 (not Mini) receiver?
![Jumper R900](https://user-images.githubusercontent.com/94545140/142225478-d28ebba5-0f9c-4f64-9101-39bf4b9e45ed.png)

It's inexpensive (~15$) comes with antenna diversity, 8 PWM channels + some more I/O pins and is based on a STM32F103C8T6 (as R900 Mini).
I'm not really experienced in that kind of work but I already did some investigations and have an almost complete picture of the hardware architecture and uC pin assignment. With some experienced guidance and support I would like to continue in this task.

![P1030011](https://user-images.githubusercontent.com/94545140/142224967-509ccbe8-e8b5-41e9-a349-02a174cfb56c.JPG)

**Comment by svpetar** [comment]
Great! I am thinking about this some time already.

**Comment by svpetar** [comment]
Compare it to Jumper R900 mini as described on: https://www.expresslrs.org/1.0/quick-start/rx-jumper900/
There is same chip and connection. I check it and found as on the picture in attachment
![Clipboard01](https://user-images.githubusercontent.com/83229067/148529883-ffe264ed-1601-4969-a4d4-1777ab7fffdd.jpg)
.

**Reply by FelixSchreiber** [reply]
Hi svpetar, 
you wrote Jumper R8 above - is this a typo? Looks like the picture of the R900 (but I also noticed quite some similarity to R8)

Regarding the programming interface - I can confirm.
Regarding the Tx1 (and also R900 usage of Tx2/Rx2) see my next post

**Comment by FelixSchreiber** [comment]
I have some more information to share

Regarding Rx/Tx (blue pin header labeled 'D' next to Ch4 **and** Ch6):
This is done as on the Jumper R8 receiver ([schematic](https://www.rcmodelytt.sk/wp-content/uploads/2020/01/Jumper_R8v2-schematics.pdf)) ![grafik](https://user-images.githubusercontent.com/94545140/148539004-c1f3ec62-4877-4f6a-9423-b2998c6bd89a.png)

![grafik](https://user-images.githubusercontent.com/94545140/148539175-8a22070b-e087-4426-b642-dc407265f6b6.png)


Regarding SBus (blue pin header labeled 'S' next to Ch3):
Connected via inverter (SN74LVC1G04 on PCB top side, next to the SX1278) to PA9 (note: Jumper R8 uses a transistor here)
![grafik](https://user-images.githubusercontent.com/94545140/148539786-34eb5e22-5bcd-4378-85d6-e4db63e07145.png)
 

And regarding the HF switches (SKY13330):
There is one for switching Rx/Tx (as on R900 mini ?) controlled via PB3 and one for switching between the two antennas controlled via PB9:
![grafik](https://user-images.githubusercontent.com/94545140/148540808-dc0991aa-03ba-4356-a260-85e356bf5c0a.png)


There is a Rx Battery monitoring on PA4 similar to Jumper R8 (not sure if there are the same mistakes regarding the diode direction and the mixed resistors for the voltage divider as on the R8):
![grafik](https://user-images.githubusercontent.com/94545140/148543154-d92dd340-a5cb-40a0-873b-9e4f299a70a3.png)

**Comment by FelixSchreiber** [comment]
This is my current picture of the STM32 peripherals:
![grafik](https://user-images.githubusercontent.com/94545140/148541517-4d88bc8c-6b43-4430-8f65-da01f5cab20e.png)
(green dots = verified; yellow = assumption; red = no idea; grey = ignored)

**Reply by svpetar** [reply]
Thank you for detailed info. I made misstake about Tx1 position on Jumper R900 receiver.
![148529883-ffe264ed-1601-4969-a4d4-1777ab7fffdd](https://user-images.githubusercontent.com/83229067/148600217-8e1bfc21-80b1-4f7c-8698-7ad96d3d32cb.jpg)

**Reply by svpetar** [reply]
> This is my current picture of the STM32 peripherals: ![grafik](https://user-images.githubusercontent.com/94545140/148541517-4d88bc8c-6b43-4430-8f65-da01f5cab20e.png) (green dots = verified; yellow = assumption; red = no idea; grey = ignored)

Hi Felix! I flashed R900 receiver similarly to R900 mini and it works on S-port. I used same firmware as R900 mini.

**Comment by svpetar** [comment]
Hi Felix! I flashed R900 receiver similarly to R900 mini and it works on S-port. I used same firmware as R900 mini. I still did not try PWM possibilities as I expected that it would work too. Unfortunately it does not. Only good thing is that R900 is much cheaper than any other receiver. Anyway I much more like an idea to use R900 wit external TX module FrSky R9M with native PWM.

**Reply by FelixSchreiber** [reply]
great news - glory goes to the brave!
I'll follow and try if I can understand the PWM outputs as the hardware peripherals are known; same would go for the antenna diversity as addon to R900 mini functionality

**Comment by FelixSchreiber** [comment]
Hi @svpetar! 
Yesterday I also managed to flash mine. Then I added the extras of the R900 like antenna diversity and PWM output and yes, it works! 
If you want to give it a try, here is my fork with the modified code: [https://github.com/FelixSchreiber/ExpressLRS/tree/Jumper_R900](https://github.com/FelixSchreiber/ExpressLRS/tree/Jumper_R900). 
To save the servo connectors for 'real' channels, I assigned the arming channel 5 to the SBus pin, although it gets inverted by hardware.
You can either have 6 channels (+ arming channel) and CRSF Rx/Tx on channel 5/6 or 8 channels (+ arming channel) and no CRSF interface (un-comment `-DCRSF_RCVR_NO_SERIAL` in user_defines.txt).
My modifications are based on release 2.5.0; I'm currently struggling to port it to the master branch because the PWM output part has been completely changed and is currently only available for ESP based platforms

**Comment by svpetar** [comment]
Great work Felix! Thank's a lot. I'll follow.

**Comment by airforce888** [comment]
THanks Felix, This is great as I have been waiting for your great work for almost a year!

**Comment by gnuton** [comment]
is this now  supported or this task is still open?
@FelixSchreiber is there anything to do here?

**Reply by FelixSchreiber** [reply]
Hi Antonio,
my fork (see above link) works for ExpressLRS 2.5. Actually the changes are not that much compared to the R900 mini but there are many options like having Crossfire Rx/Tx on some channel pins or even with minor HW modifications to use other pins for Rx/Tx.
I was not able to port it to 3.0 because in 3.0 the PWM subsystem was completely rewritten exclusively for Espressif controllers. I started with an idea to use the new 3.0 PWM interface for the old Arduino servo library, but I think my approach is dirty and I don't fully understand the new PWM subsystem - so this is something to be done...
Best regards,
 Felix

**Reply by gnuton** [reply]
Thanks a lot for the reply @FelixSchreiber

**Comment by amatafix** [comment]
I would like to know where is the progress of jumper r900 elrs now, can it be used with elrs 3.0 or not?

**Reply by FelixSchreiber** [reply]
no change since Jan 26 :-(

But it would be great if all who are interested can motivate some more experienced devs to fix the missing pieces

**Comment by svpetar** [comment]
Agree!

    On Monday, May 1, 2023 at 12:03:26 PM GMT+2, FelixSchreiber ***@***.***> wrote:  
 
 


no change since Jan 26 :-(

But it would be great if all who are interested can motivate some more experienced devs to fix the missing pieces

**Comment by FelixSchreiber** [comment]
@svpetar, @airforce888, @gnuton  @amatafix - I started a new attempt to port R900 support to recent master branch:
https://github.com/FelixSchreiber/ExpressLRS/tree/support-jumper-r900-with-pwm

For now it compiles but I didn't test it so far. So if you are brave, give it a try. 
If you have some coding knowledge, please review and help me improving the code!!!

**Comment by jjsueldo** [comment]
Hello. Searching and searching I came here because I want to change the original firmware of the RX Jumper R900. I'm moving from FrSky to ExLRS and I don't have accurate information on how to flash this thing safely with an FTDI
Does anyone know how to wire the FTDI and the R900?

**Reply by svpetar** [reply]
Hi! I flash it as it is explained above, but there is no PWM exit to servos as we want it. Felix also did not finish try. Maybe is now time as there is finally ELRS community made receivers with PWM.
